cpu/stm32_common/periph/dma.c

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright (C) 2017 OTA keys S.A.
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_stm32_common
+ * @{
+ *
+ * @file
+ * @brief       Low-level DMA driver implementation
+ *
+ * @author      Vincent Dupont <vincent@otakeys.com>
+ *
+ * @}
+ */
+
+#include <stdint.h>
+
+#include "periph_cpu.h"
+#include "periph_conf.h"
+#include "mutex.h"
+#include "assert.h"
+
+#if defined(DMA_NUMOF) && (defined(CPU_FAM_STM32F2) || defined(CPU_FAM_STM32F4) || defined(CPU_FAM_STM32F7))
+
+#define DMA_STREAM_IT_MASK   (DMA_LISR_FEIF0 | DMA_LISR_DMEIF0 | \
+                              DMA_LISR_TEIF0 | DMA_LISR_HTIF0 | \
+                              DMA_LISR_TCIF0)
+
+struct dma_ctx {
+    mutex_t conf_lock;
+    mutex_t sync_lock;
+    uint16_t len;
+};
+
+static struct dma_ctx dma_ctx[DMA_NUMOF];
+
+static inline uint32_t dma_all_flags(dma_t dma)
+{
+    assert(dma < DMA_NUMOF);
+
+    switch (dma_config[dma].stream & 0x3) {
+    case 0: /* 0 and 4 */
+        return (DMA_STREAM_IT_MASK);
+    case 1: /* 1 and 5 */
+        return (DMA_STREAM_IT_MASK << 6);
+    case 2: /* 2 and 6 */
+        return (DMA_STREAM_IT_MASK << 16);
+    case 3: /* 3 and 7 */
+        return (DMA_STREAM_IT_MASK << 22);
+    default:
+        return 0;
+    }
+}
+
+static void dma_clear_all_flags(dma_t dma)
+{
+    assert(dma < DMA_NUMOF);
+
+    DMA_TypeDef *stream = dma_base(dma_config[dma].stream);
+
+    /* Clear all flags */
+    if (dma_hl(dma_config[dma].stream) == 0) {
+        stream->LIFCR = dma_all_flags(dma);
+    }
+    else {
+        stream->HIFCR = dma_all_flags(dma);
+    }
+}
+
+
+void dma_init(void)
+{
+    for (int i = 0; i < DMA_NUMOF; i++) {
+        mutex_init(&dma_ctx[i].conf_lock);
+        mutex_init(&dma_ctx[i].sync_lock);
+        mutex_lock(&dma_ctx[i].sync_lock);
+    }
+}
+
+int dma_transfer(dma_t dma, int chan, const void *src, void *dst, size_t len, dma_mode_t mode, uint8_t flags)
+{
+    dma_acquire(dma);
+    int ret = dma_configure(dma, chan, src, dst, len, mode, flags);
+    if (ret != 0) {
+        return ret;
+    }
+    dma_start(dma);
+    dma_wait_for_end(dma);
+    dma_stop(dma);
+    dma_release(dma);
+
+    return len;
+}
+
+void dma_acquire(dma_t dma)
+{
+    assert(dma < DMA_NUMOF);
+
+    mutex_lock(&dma_ctx[dma].conf_lock);
+}
+
+void dma_release(dma_t dma)
+{
+    assert(dma < DMA_NUMOF);
+
+    mutex_unlock(&dma_ctx[dma].conf_lock);
+}
+
+int dma_configure(dma_t dma, int chan, const void *src, void *dst, size_t len, dma_mode_t mode, uint8_t flags)
+{
+    assert(src != NULL);
+    assert(dst != NULL);
+    assert(dma < DMA_NUMOF);
+
+    int stream_n = dma_config[dma].stream;
+    uint32_t inc_periph;
+    uint32_t inc_mem;
```

Comment:
~~The increments can be deduced from the mode, right? M2M, (`PINC | MINC`) and M2P, P2M (`MINC`). No need to have them in flags. So always `MINC` and only M2M `|= PINC`.~~

Now I see the trick in the spi implementation. Forget about it.